src/jetstream_kv.rs

```diff
@@ -0,0 +1,906 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use nats::jetstream::KeyValueConfig;
+    /// # fn main() -> std::io::Result<()> {
+    /// # let client = nats::connect("demo.nats.io")?;
+    /// # let context = nats::jetstream::new(client);
+    /// #
+    /// # let bucket = context.create_key_value(&KeyValueConfig {
+    /// #  bucket: "get".to_string(),
+    /// #  ..Default::default()
+    /// # })?;
+    /// #
+    /// bucket.put("foo", b"bar")?;
+    /// let maybe_value = bucket.get("foo")?;
+    /// if let Some(value) = maybe_value {
+    ///   println!("Found value {:?}", value);
+    /// }
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn get(&self, key: &str) -> io::Result<Option<Vec<u8>>> {
+        match self.entry(key) {
+            Ok(Some(entry)) => match entry.operation {
+                KeyValueOperation::Put => Ok(Some(entry.value)),
+                _ => Ok(None),
+            },
+            Ok(None) => Ok(None),
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Places the new value for the key into the bucket.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use nats::jetstream::KeyValueConfig;
+    /// # fn main() -> std::io::Result<()> {
+    /// # let client = nats::connect("demo.nats.io")?;
+    /// # let context = nats::jetstream::new(client);
+    /// #
+    /// # let bucket = context.create_key_value(&KeyValueConfig {
+    /// #  bucket: "get".to_string(),
+    /// #  ..Default::default()
+    /// # })?;
+    /// #
+    /// bucket.put("foo", b"bar")?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn put(&self, key: &str, value: impl AsRef<[u8]>) -> io::Result<u64> {
+        if !is_valid_key(key) {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput, "invalid key"));
+        }
+
+        let mut subject = String::new();
+        subject.push_str(&self.prefix);
+        subject.push_str(key);
+
+        let publish_ack = self.context.publish(&subject, value)?;
+
+        Ok(publish_ack.sequence)
+    }
+
+    /// Creates the key/value pair if it does not exist.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use nats::jetstream::KeyValueConfig;
+    /// # fn main() -> std::io::Result<()> {
+    /// # let client = nats::connect("demo.nats.io")?;
+    /// # let context = nats::jetstream::new(client);
+    /// #
+    /// # let bucket = context.create_key_value(&KeyValueConfig {
+    /// #  bucket: "create".to_string(),
+    /// #  ..Default::default()
+    /// # })?;
+    /// #
+    /// bucket.purge("foo")?;
+    /// bucket.create("foo", b"bar")?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn create(&self, key: &str, value: impl AsRef<[u8]>) -> io::Result<u64> {
+        if !is_valid_key(key) {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput, "invalid key"));
+        }
+
+        let revision = if let Ok(Some(entry)) = self.entry(key) {
```

Comment:
Setting revision to `0` if `self.entry(key)` is Err.

I guess it was not intentional and was meant only for `Some`, right?

Reply:
That's intentional, I'll flatten this flow a bit to be more along the happy path.